README.md
# cool-helpers 
[![Build Status](https://secure.travis-ci.org/bq/cool-helpers.png?branch=master)](http://travis-ci.org/bq/cool-helpers)

> Cool Helpers is a Handlebars.js helpers compilation with AMD support

## Jump to Section

* [Quickstart](#quickstart)
* [Usage](#usage)
* [Commons](#commons)
* [Comparison](#comparison)
* [Math](#math)
* [Date](#date)

## Quickstart
[[Back To Top]](#jump-to-section)

To start using cool-helpers you can:

* [Download the last release](https://github.com/bq/cool-helpers/releases/tag/v0.0.2)
* [Install with bower](http://bower.io/): `bower install cool-helpers --save-dev`

###Build the project

If you want to build the project locally, you have to run the [Grunt](http://gruntjs.com/) task `grunt build`. This task will concat and minify the sources and create the README.md. Before all this, you do not forget to run `npm install` and `bower install`. 

####Contribute

If you want to keep growing cool-helpers up, send us a pull request with your new helpers. You must follow this steps:

* Add the helper to its proper category file inside src/helpers directory. If you want to create a new category, create the new js file and add it at src/helpers/start.js.
* Do your commit message following the [commit conventions](https://github.com/ajoslin/conventional-changelog/blob/master/CONVENTIONS.md). The changelog will be created using the commited messages.
* Document the helper. No pull request will be accepted without proper documentation. Documentation directory is doc. If you need to create a new documentation section, remember to add it at Gruntfile, inside readme_generator.readme.order. Also, you have to write low-level documentation using JSDoc.
* Remember to build the project before doing the pull-request. 


## Usage
[[Back To Top]](#jump-to-section)

    'use strict'

    require.config({
      deps: [
        'cool-helpers'
      ],
      paths: {
        'cool-helpers': 'bower_components/cool-helpers/lib/cool-helpers',
        handlebars: 'bower_components/handlebars/handlebars',
        moment: 'bower_components/moment/moment'
      }      
    });


## Commons
[[Back To Top]](#jump-to-section)

> Commons

###data

> Generates data-attribute from context.type and context.value

    {{#data attribute}}{{/data}}

###datas

> Generates data-attributes from context.type and context.value

    {{#datas attributes}}{{/datas}}

###attributes

> Generates data-attribute from context.type and context.value

    {{#attributes attributes}}{{/attributes}}

###macro

> Generates a new helper with name key

    {{#macro "mymacro" options=""}}
        {{#each options}}
        {{/each}}
    {{/macro}}
     
    {{mymacro options=options}

###nl2br

> Changes each nl to a br tag

    {{#nl2br title}}{{/nl2br}}

###log

> Log one or multiple values to the console:

    {{#log title}}{{/log}}

###debug

> Log one or multiple values to the console, with the current context:

    {{#debug title}}{{/debug}}


## Comparison
[[Back To Top]](#jump-to-section)

> Comparison

###contains

Conditionally render a block if string matches pattern

    {{#contains title "pattern"}}
        text
    {{/contains}}

###and

Conditionally render a block if both a and b properties exists and are not null or undefined

    {{#and title name}}
        text
    {{/and}}

###gt

Conditionally render a block if value > test

    {{#@gt 5 2}}
        text
    {/gt}}

###gte

Conditionally render a block if value >= test

    {{#@gte 5 2}}
        text
    {/gte}}

###is

Conditionally render a block if value === test

    {{#@is 5 5}}
        text
    {/is}}

###isnt

Conditionally render a block if value !== test

    {{#@isnt 5 5}}
        text
    {/isnt}}

###lt

Conditionally render a block if value < test

    {{#@lt 5 2}}
        text
    {/lt}}

###lte

Conditionally render a block if value <= test

    {{#@lte 5 2}}
        text
    {/lte}}

###or

Conditionally render a block if a || b

    {{#@or a b}}
        text
    {/or}}

###ifNth

Conditionally render a block if nr%v === 0

    {{#ifNth 10 2}}
        text
    {{/ifNth}}

###compare 

Compares two values with given operator. Operators are: ==, ===, !=, !==, <, >, <=, >=, typeof. Operator must be between quotes "";

    {{#compare unicorns "<" ponies}}
        I knew it, unicorns are just low-quality ponies!
    {{/compare}}

    {{#compare value ">=" 10}}
        The value is greater or equal than 10
    {{else}}
        The value is lower than 10
    {{/compare}}















## Math
[[Back To Top]](#jump-to-section)

> Math

###add

> Returns value + adition

    {{add 5 1}}

###substract

> Returns value - substraction

    {{substract 5 1}}

###divide

> Returns value % divisor

    {{divide 6 3}}

###multiply

> Returns value * multiplier

    {{multiply 5 2}}

###floor

Returns Math.floor(value)

    {{floor 3.2222}}

###ceil

Returns Math.ceil(value)

    {{ceil 3.2222}}

###round

Returns Math.round(value)

    {{round 3.2222}}

###sum

Returns sum of each argument

    {{sum 1 1 2 3 5 8}}






## Date
[[Back To Top]](#jump-to-section)

> Date

###formatDate

> Formats date in given format

    {{formatDate date 'yyyy/mm/dd'}}

###subtractDate

> Returns the difference between two dates in given format

    {{subtractDates date1 date2 "d"}}





--------
<small>_This readme has been automatically generated by [readme generator](https://github.com/aponxi/grunt-readme-generator) on Tue Mar 10 2015 10:19:07 GMT+0100 (CET)._</small>

CHANGELOG.md
<a name="0.0.2"></a>
### 0.0.2 (2015-03-04)


test/spec/math.js
describe('Math helpers', function() {

    describe('Sum', function() {
        it('should return "2" when we sum 1 + 1', function() {
            var fields = {
                    a: 1,
                    b: 1
                },
                html = "{{add a b}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('2');
        });
    });

    describe('Subtract', function() {
        it('should return "1" when we subtract 2 - 1', function() {
            var fields = {
                    a: 2,
                    b: 1
                },
                html = "{{subtract a b}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('1');
        });
    });

    describe('Divide', function() {
        it('should return "2" when we divide 4 - 2', function() {
            var fields = {
                    a: 4,
                    b: 2
                },
                html = "{{divide a b}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('2');
        });
    });

    describe('Multiply', function() {
        it('should return "4" when we multiply 2 * 2', function() {
            var fields = {
                    a: 2,
                    b: 2
                },
                html = "{{multiply a b}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('4');
        });
    });

    describe('Floor', function() {
        it('should return "2" when we floor 2.5', function() {
            var fields = {
                    a: 2.5
                },
                html = "{{floor a}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('2');
        });
    });

    describe('Ceil', function() {
        it('should return "3" when we ceil 2.1', function() {
            var fields = {
                    a: 2.1
                },
                html = "{{ceil a}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('3');
        });
    });

    describe('Round', function() {
        it('should return "3" when we ceil 2.5', function() {
            var fields = {
                    a: 2.5
                },
                html = "{{round a}}",
                template = Handlebars.compile(html);

            expect(template(fields)).to.equal('3');
        });
    });

});


test/spec/setup.js
before(function() {
    Handlebars = require("handlebars"),
    coolHelpers = require("../../lib/cool-helpers"),
    chai = require('chai'),
    helpers = Handlebars.helpers,
    expect = chai.expect;
});


Gruntfile.js
/* global module, require */

'use strict';

module.exports = function(grunt) {

    require('load-grunt-tasks')(grunt);

    grunt.initConfig({
        bw: grunt.file.readJSON('bower.json'),
        pkg: grunt.file.readJSON('package.json'),
        readme_generator: {
            readme: {
                options: {
                    github_username: 'bq',
                    readme_folder: 'doc'
                },
                order: {
                    'quickstart.md': 'Quickstart',
                    'usage.md': 'Usage',
                    'commons.md': 'Commons',
                    'comparison.md': 'Comparison',
                    'math.md': 'Math',
                    'date.md': 'Date'
                }
            }
        },
        preprocess: {
            helpers: {
                src: 'src/start.js',
                dest: 'lib/cool-helpers.js'
            }
        },
        mochacli: {
            options: {
                bail: true
            },
            all: ['test/spec/*.js']
        },
        uglify: {
            dist: {
                files: {
                    'lib/cool-helpers.min.js': ['lib/cool-helpers.js']
                }
            }
        },
        concat: {
            options: {
                stripBanners: true,
                banner: '/*! <%= bw.name %> - v<%= bw.version %> - @author: bq.com -' + '<%= grunt.template.today("yyyy-mm-dd") %> */\n'
            },
            version: {
                src: ['lib/cool-helpers.js'],
                dest: 'lib/cool-helpers.js'
            }
        },
        bump: {
            options: {
                files: ['bower.json'],
                updateConfigs: ['bw'],
                commit: true,
                commitMessage: 'Release v%VERSION%',
                commitFiles: ['bower.json', 'CHANGELOG.md'],
                createTag: true,
                tagName: 'v%VERSION%',
                tagMessage: 'Version %VERSION%',
                push: true,
                pushTo: '<%= bw.repository.url %>',
                gitDescribeOptions: '--tags --always --abbrev=1 --dirty=-d',
                globalReplace: false,
                prereleaseName: false,
                regExp: false
            }
        },
        changelog: {
            options: {
                version: '<%= bw.version %>',
                repository: '<%= bw.repository.url %>'
            }
        },
    });

    grunt.registerTask('_commit', [
        'changelog',
        'bump-commit'
    ]);

    grunt.registerTask('build', [
        'preprocess:helpers',
        'concat',
        'uglify',
        'readme_generator'
    ]);

    grunt.registerTask('test', [
        'mochacli'
    ]);

    grunt.registerTask('release', function(version) {
        var semVer = /\bv?(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)(?:-[\da-z\-]+(?:\.[\da-z\-]+)*)?(?:\+[\da-z\-]+(?:\.[\da-z\-]+)*)?\b/ig;

        if (!semVer.test(version)) {
            grunt.option('setversion', false);
            grunt.task.run('bump-only:' + version);
        } else {
            grunt.option('setversion', version);
            grunt.task.run('bump-only');
        }

        grunt.task.run('_commit');
    });
};


src/start.js
(function(root, factory) {

    if (typeof define === 'function' && define.amd) {
        define(['handlebars', 'moment'], function(Handlebars, moment) {
            return (root.CoolHelpers = root.CoolH = factory(root, Handlebars, moment));
        });
    } else if (typeof exports !== 'undefined') {
        var Handlebars = require('handlebars'),
            moment = require('moment');
        module.exports = factory(root, Handlebars, moment);
    } else {
        root.CoolHelpers = root.CoolH = factory(root, root.Handlebars, root.moment);
    }

}(this, function(root, Handlebars, moment) {

    'use strict';

    // @include core.js

    // @include helpers/common.js

    // @include helpers/comparison.js

    // @include helpers/date.js

    // @include helpers/math.js

    helpers.register(
        common,
        comparison,
        date,
        math
    );

    return helpers;

}));


src/helpers/date.js
/**
 * Date helpers set
 * @namespace date
 */
var date = {
    /**
     * Returns given date with given format
     * @memberof date
     * @param  {string} date      Date to format
     * @param  {string} format    Date format
     * @param  {string} errorMsg  Error message key
     *
     * @example
     *
     * {{formatDate date 'yyyy/mm/dd'}}
     */
    formatDate: function(date, format) {
        return moment(date).format(format);
    },

    /**
     * Returns the difference between two dates in given format
     * @memerof date
     * @param  {string} startDate   Start date
     * @param  {string} endDate     End date
     * @param  {string} format      Date difference format
     * @return {string}             formated date
     */
    subtractDates: function(startDate, endDate, format) {
        return moment(moment(moment(startDate)).diff(moment(endDate))).format(format);
    },

};


src/helpers/math.js
/**
 * Math helpers set
 * @namespace math
 */
var math = {

    /**
     * Returns value + addition
     * @memberof math
     * @param {number} value    [description]
     * @param {number} addition [description]
     *
     * @example
     *
     * {{add 5 1}}
     */
    add: function(value, addition) {
        return value + addition;
    },

    /**
     * Returns value - substraction
     * @memberof math
     * @param {number} value    [description]
     * @param {number} substraction [description]
     *
     * @example
     *
     * {{substract 5 1}}
     */
    subtract: function(value, substraction) {
        return value - substraction;
    },

    /**
     * Returns value % divisor
     * @memberof math
     * @param {number} value    [description]
     * @param {number} divisor [description]
     *
     * @example
     *
     * {{divide 6 3}}
     */
    divide: function(value, divisor) {
        return value / divisor;
    },

    /**
     * Returns value * multiplier
     * @memberof math
     * @param {number} value    [description]
     * @param {number} multiplier [description]
     *
     * @example
     *
     * {{multiply 3 2}}
     */
    multiply: function(value, multiplier) {
        return value * multiplier;
    },

    /**
     * Returns Math.floor(value)
     * @memberof math
     * @param {number} value    [description]
     *
     * @example
     *
     * {{floor 3.2222}}
     */
    floor: function(value) {
        return Math.floor(value);
    },

    /**
     * Returns Math.ceil(value)
     * @memberof math
     * @param {number} value    [description]
     *
     * {{ceil 3.2222}}
     */
    ceil: function(value) {
        return Math.ceil(value);
    },

    /**
     * Returns Math.round(value)
     * @memberof math
     * @param {number} value    [description]
     *
     * @example
     *
     * {{round 3.222}}
     */
    round: function(value) {
        return Math.round(value);
    }
};


src/helpers/common.js
/**
 * Common helpers set
 * @namespace comparison
 */
var common = {

    /**
     * Generates a new helper with name key
     * @memberof common
     * @param  {string} name Helper name
     *
     * @example
     *
     * {{#macro "mymacro" options=""}}
     *     {{#each options}}
     *     {{/each}}
     * {{/macro}}
     *
     * {{mymacro options=options}
     *
     */
    macro: function(name, defaults) {
        Handlebars.registerHelper(name, function(options) {
            var e = $.extend(this, defaults.hash, options.hash);
            return new Handlebars.SafeString(defaults.fn(e));
        });
    },

    /**
     * Changes each nl to a br tag
     * @memberof common
     * @param  {string} text Text to apply nl2br
     *
     * @example
     * {{#nl2br title}}{{/nl2br}}
     */
    nl2br: function(text) {
        var nl2br = (text + '').replace(/([^>\r\n]?)(\r\n|\n\r|\r|\n)/g, '$1' + '<br>' + '$2');
        return new Handlebars.SafeString(nl2br);
    },

    /**
     * Log one or multiple values to the console:
     * @memberof common
     *
     * @example
     * {{#log title}}{{/log}}
     */
    log: function() {
        console.log(['Values:'].concat(
            Array.prototype.slice.call(arguments, 0, -1)
        ));
    },

    /**
     * Log one or multiple values to the console, with the current context:
     * {{debug foo bar}
     * @memberof common
     *
     * @example
     * {{#debug title}}{{/debug}}
     */
    debug: function() {
        console.log('Context:', this);
        console.log(['Values:'].concat(
            Array.prototype.slice.call(arguments, 0, -1)
        ));
    }

};


lib/cool-helpers.js
/*! cool-helpers - v0.0.2 - @author: bq.com -2015-03-10 */
(function(root, factory) {

    if (typeof define === 'function' && define.amd) {
        define(['handlebars', 'moment'], function(Handlebars, moment) {
            return (root.CoolHelpers = root.CoolH = factory(root, Handlebars, moment));
        });
    } else if (typeof exports !== 'undefined') {
        var Handlebars = require('handlebars'),
            moment = require('moment');
        module.exports = factory(root, Handlebars, moment);
    } else {
        root.CoolHelpers = root.CoolH = factory(root, root.Handlebars, root.moment);
    }

}(this, function(root, Handlebars, moment) {

    'use strict';

    var helpers = {
        register: function() {
            for (var i = 0, length = arguments.length; i < length; i++) {
                for (var helper in arguments[i]) {
                    if (arguments[i].hasOwnProperty(helper)) {
                        Handlebars.registerHelper(helper, arguments[i][helper]);
                    }
                }
            };
        }
    };
    

    /**
     * Common helpers set
     * @namespace comparison
     */
    var common = {
    
        /**
         * Generates a new helper with name key
         * @memberof common
         * @param  {string} name Helper name
         *
         * @example
         *
         * {{#macro "mymacro" options=""}}
         *     {{#each options}}
         *     {{/each}}
         * {{/macro}}
         *
         * {{mymacro options=options}
         *
         */
        macro: function(name, defaults) {
            Handlebars.registerHelper(name, function(options) {
                var e = $.extend(this, defaults.hash, options.hash);
                return new Handlebars.SafeString(defaults.fn(e));
            });
        },
    
        /**
         * Changes each nl to a br tag
         * @memberof common
         * @param  {string} text Text to apply nl2br
         *
         * @example
         * {{#nl2br title}}{{/nl2br}}
         */
        nl2br: function(text) {
            var nl2br = (text + '').replace(/([^>\r\n]?)(\r\n|\n\r|\r|\n)/g, '$1' + '<br>' + '$2');
            return new Handlebars.SafeString(nl2br);
        },
    
        /**
         * Log one or multiple values to the console:
         * @memberof common
         *
         * @example
         * {{#log title}}{{/log}}
         */
        log: function() {
            console.log(['Values:'].concat(
                Array.prototype.slice.call(arguments, 0, -1)
            ));
        },
    
        /**
         * Log one or multiple values to the console, with the current context:
         * {{debug foo bar}
         * @memberof common
         *
         * @example
         * {{#debug title}}{{/debug}}
         */
        debug: function() {
            console.log('Context:', this);
            console.log(['Values:'].concat(
                Array.prototype.slice.call(arguments, 0, -1)
            ));
        }
    
    };
    

    /**
     * Comparsion helpers set
     * @namespace comparison
     */
    var comparison = {
        /**
         * Conditionally render a block if string matches pattern
         * @memberof comparison
         * @param  {string} str     Value to be compared with pattern
         * @param  {string} pattern Patter which value will be compared to
         *
         * @example
         *
         * {{#contains title "pattern"}}
         *     text
         * {{/contains}}
         */
        contains: function(str, pattern, options) {
            if (str.indexOf(pattern) !== -1) {
                return options.fn(this);
            }
            return options.inverse(this);
        },
    
        /**
         * Conditionally render a block if both a and b properties exists and are not null or undefined
         * @memberof comparison
         * @param  {primitive} a first property
         * @param  {primitive} b second property
         *
         * @example
         *
         * {{#and title name}}
         *     text
         * {{/and}}
         */
        and: function(a, b, options) {
            if (a && b) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if value > test
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#@gt 5 2}}
         *     text
         * {{/gt}}
         */
        gt: function(value, test, options) {
            if (value > test) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if value >= test
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#gte 5 2}}
         *     text
         * {{/gte}}
         */
        gte: function(value, test, options) {
            if (value >= test) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if value === test
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#is name "John Doe"}}
         *     text
         * {{/is}}
         */
        is: function(value, test, options) {
            if (value === test) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if value !== test
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#is name "John Doe"}}
         *     text
         * {{/is}}
         */
        isnt: function(value, test, options) {
            if (value !== test) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if value < test
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#lt 2 5}}
         *     text
         * {{/lt}}
         */
        lt: function(value, test, options) {
            if (value < test) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if value <= test
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#lt 2 5}}
         *     text
         * {{/lt}}
         */
        lte: function(value, test, options) {
            if (value <= test) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if a || b
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test element with first value will be compared to
         *
         * @example
         *
         * {{#or name surname}}
         *     text
         * {{/or}}
         */
        or: function(a, b, options) {
            if (a || b) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * Conditionally render a block if nr%v === 0
         * @memberof comparison
         * @param  {primitive} value value to test
         * @param  {primitive} test eleme
         *
         * @example
         *
         * {{#ifNth 10 2}}
         *     text
         * {{/ifNth}}
         */
        ifNth: function(nr, v, options) {
            v = v + 1;
            if (v % nr === 0) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        },
    
        /**
         * {{#compare}}...{{/compare}}
         *
         * @memberof comparison
         *
         * @credit: OOCSS
         * @param left value
         * @param operator The operator, must be between quotes ">", "=", "<=", etc...
         * @param right value
         * @param options option object sent by handlebars
         * @return {String} formatted html
         *
         * @example
         *   {{#compare unicorns "<" ponies}}
         *     I knew it, unicorns are just low-quality ponies!
         *   {{/compare}}
         *
         *   {{#compare value ">=" 10}}
         *     The value is greater or equal than 10
         *     {{else}}
         *     The value is lower than 10
         *   {{/compare}}
         */
        compare: function(left, operator, right, options) {
            /*jshint eqeqeq: false*/
    
            if (arguments.length < 3) {
                throw new Error('Handlebars Helper "compare" needs 2 parameters');
            }
    
            if (options === undefined) {
                options = right;
                right = operator;
                operator = '===';
            }
    
            var operators = {
                '==': function(l, r) {
                    return l == r;
                },
                '===': function(l, r) {
                    return l === r;
                },
                '!=': function(l, r) {
                    return l != r;
                },
                '!==': function(l, r) {
                    return l !== r;
                },
                '<': function(l, r) {
                    return l < r;
                },
                '>': function(l, r) {
                    return l > r;
                },
                '<=': function(l, r) {
                    return l <= r;
                },
                '>=': function(l, r) {
                    return l >= r;
                },
                'typeof': function(l, r) {
                    return typeof l == r;
                }
            };
    
            if (!operators[operator]) {
                throw new Error('Handlebars Helper "compare" doesn\'t know the operator ' + operator);
            }
    
            var result = operators[operator](left, right);
    
            if (result) {
                return options.fn(this);
            } else {
                return options.inverse(this);
            }
        }
    };
    

    /**
     * Date helpers set
     * @namespace date
     */
    var date = {
        /**
         * Returns given date with given format
         * @memberof date
         * @param  {string} date      Date to format
         * @param  {string} format    Date format
         * @param  {string} errorMsg  Error message key
         *
         * @example
         *
         * {{formatDate date 'yyyy/mm/dd'}}
         */
        formatDate: function(date, format) {
            return moment(date).format(format);
        },
    
        /**
         * Returns the difference between two dates in given format
         * @memerof date
         * @param  {string} startDate   Start date
         * @param  {string} endDate     End date
         * @param  {string} format      Date difference format
         * @return {string}             formated date
         */
        subtractDates: function(startDate, endDate, format) {
            return moment(moment(moment(startDate)).diff(moment(endDate))).format(format);
        },
    
    };
    

    /**
     * Math helpers set
     * @namespace math
     */
    var math = {
    
        /**
         * Returns value + addition
         * @memberof math
         * @param {number} value    [description]
         * @param {number} addition [description]
         *
         * @example
         *
         * {{add 5 1}}
         */
        add: function(value, addition) {
            return value + addition;
        },
    
        /**
         * Returns value - substraction
         * @memberof math
         * @param {number} value    [description]
         * @param {number} substraction [description]
         *
         * @example
         *
         * {{substract 5 1}}
         */
        subtract: function(value, substraction) {
            return value - substraction;
        },
    
        /**
         * Returns value % divisor
         * @memberof math
         * @param {number} value    [description]
         * @param {number} divisor [description]
         *
         * @example
         *
         * {{divide 6 3}}
         */
        divide: function(value, divisor) {
            return value / divisor;
        },
    
        /**
         * Returns value * multiplier
         * @memberof math
         * @param {number} value    [description]
         * @param {number} multiplier [description]
         *
         * @example
         *
         * {{multiply 3 2}}
         */
        multiply: function(value, multiplier) {
            return value * multiplier;
        },
    
        /**
         * Returns Math.floor(value)
         * @memberof math
         * @param {number} value    [description]
         *
         * @example
         *
         * {{floor 3.2222}}
         */
        floor: function(value) {
            return Math.floor(value);
        },
    
        /**
         * Returns Math.ceil(value)
         * @memberof math
         * @param {number} value    [description]
         *
         * {{ceil 3.2222}}
         */
        ceil: function(value) {
            return Math.ceil(value);
        },
    
        /**
         * Returns Math.round(value)
         * @memberof math
         * @param {number} value    [description]
         *
         * @example
         *
         * {{round 3.222}}
         */
        round: function(value) {
            return Math.round(value);
        }
    };
    

    helpers.register(
        common,
        comparison,
        date,
        math
    );

    return helpers;

}));
